File: waterui-ffi/src/component/slider.rs
use core::ops::RangeInclusive;

use waterui::component::form::slider::SliderConfig;

use crate::{binding::waterui_binding_double, waterui_anyview, IntoFFI};

#[repr(C)]
pub struct waterui_slider {
    label: *mut waterui_anyview,
    min_value_label: *mut waterui_anyview,
    max_value_label: *mut waterui_anyview,
    range: waterui_range_inclusive<f64>,
    value: *mut waterui_binding_double,
}

#[repr(C)]
pub struct waterui_range_inclusive<T> {
    start: T,
    end: T,
}

impl<T: Clone + IntoFFI> IntoFFI for RangeInclusive<T> {
    type FFI = waterui_range_inclusive<T::FFI>;
    fn into_ffi(self) -> Self::FFI {
        Self::FFI {
            start: self.start().clone().into_ffi(),
            end: self.end().clone().into_ffi(),
        }
    }
}

into_ffi!(
    SliderConfig,
    waterui_slider,
    label,
    min_value_label,
    max_value_label,
    range,
    value
);

native_view!(
    SliderConfig,
    waterui_slider,
    waterui_view_force_as_slider,
    waterui_view_slider_id
);

File: Sources/WaterUI/Slider.swift
//
//  Slider.swift
//  waterui-swift
//
//  Created by Lexo Liu on 10/21/24.
//

import SwiftUI
import CWaterUI
struct Slider:View,Component{
    static var id=waterui_view_slider_id()
    var label:WaterUI.AnyView
    var min_value_label: WaterUI.AnyView
    var max_value_label: WaterUI.AnyView
    var range: waterui_range_inclusive_f64
    @ObservedObject var value: BindingDouble
    var body: some View {
        SwiftUI.Slider(value: value.value, in: range.start...range.end, label: {
            label
        }, minimumValueLabel: {
            min_value_label
        }, maximumValueLabel: {
            max_value_label
        })
    }
    
    init(slider: waterui_slider,env:WaterUI.Environment){
        self.label=AnyView(anyview: slider.label, env: env)
        self.min_value_label=AnyView(anyview: slider.min_value_label, env: env)
        self.max_value_label=AnyView(anyview: slider.max_value_label, env: env)
        self.range=slider.range
        self.value=BindingDouble(inner: slider.value)
    }
    
    init(anyview: OpaquePointer,env:WaterUI.Environment){
        self.init(slider: waterui_view_force_as_slider(anyview), env: env)
    }
    
}
